modules/kargoBidAdapter.js

```diff
@@ -4,286 +4,484 @@ import { registerBidder } from '../src/adapters/bidderFactory.js';
+    return ++requestCounter;
+  }
+  lastPageUrl = window.location.pathname;
+  return requestCounter = 0;
+}
+
+function sendTimeoutData(auctionId, auctionTimeout) {
+  let params = {
+    aid: auctionId,
+    ato: auctionTimeout,
+  };
+
+  try {
+    let timeoutRequestUrl = buildUrl({
+      protocol: 'https',
+      hostname: 'krk.kargo.com',
+      pathname: '/api/v1/event/timeout',
+      search: params
+    });
+
+    triggerPixel(timeoutRequestUrl);
+  } catch (e) {}
+}
+
+function getImpression(bid) {
+  const imp = {
+    id: bid.bidId,
+    tid: bid.transactionId,
+    pid: bid.params.placementId,
+    code: bid.adUnitCode
+  }
+
+  if (bid.floorData != null && bid.floorData.floorMin > 0) {
+    imp.floor = bid.floorData.floorMin
+  }
+
+  if (bid.bidRequestsCount > 0) {
+    imp.bidRequestCount = bid.bidRequestsCount
+  }
+
+  if (bid.bidderRequestsCount > 0) {
+    imp.bidderRequestCount = bid.bidderRequestsCount
+  }
+
+  if (bid.bidderWinsCount > 0) {
+    imp.bidderWinCount = bid.bidderWinsCount
+  }
+
+  const gpid = getGPID(bid)
+  if (gpid != null && gpid != '') {
+    imp.fpd = {
+      gpid: gpid
+    }
+  }
+
+  if(bid.mediaTypes != null) {
+    if (bid.mediaTypes.banner != null) {
+      imp.banner = bid.mediaTypes.banner
     }
-    return userIds;
-  },
-
-  _getClientId() {
-    const crb = spec._getCrb();
-    return crb.clientId;
-  },
-
-  _getAllMetadata(bidderRequest, tdid) {
-    return {
-      userIDs: spec._getUserIds(tdid, bidderRequest.uspConsent, bidderRequest.gdprConsent),
-      pageURL: bidderRequest?.refererInfo?.page,
-      rawCRB: storage.getCookie('krg_crb'),
-      rawCRBLocalStorage: spec._getLocalStorageSafely('krg_crb')
-    };
-  },
 
-  _getSessionId() {
-    if (!sessionId) {
-      sessionId = spec._generateRandomUuid();
+    if (bid.mediaTypes.video != null) {
+      imp.video = bid.mediaTypes.video
     }
-    return sessionId;
-  },
 
-  _getRequestCount() {
-    if (lastPageUrl === window.location.pathname) {
-      return ++requestCounter;
+    if (bid.mediaTypes.native != null) {
+      imp.native = bid.mediaTypes.native
     }
-    lastPageUrl = window.location.pathname;
-    return requestCounter = 0;
-  },
-
-  _generateRandomUuid() {
-    try {
-      // crypto.getRandomValues is supported everywhere but Opera Mini for years
-      var buffer = new Uint8Array(16);
-      crypto.getRandomValues(buffer);
-      buffer[6] = (buffer[6] & ~176) | 64;
-      buffer[8] = (buffer[8] & ~64) | 128;
-      var hex = Array.prototype.map.call(new Uint8Array(buffer), function(x) {
-        return ('00' + x.toString(16)).slice(-2);
-      }).join('');
-      return hex.slice(0, 8) + '-' + hex.slice(8, 12) + '-' + hex.slice(12, 16) + '-' + hex.slice(16, 20) + '-' + hex.slice(20);
-    } catch (e) {
-      return '';
+  }
+
+  return imp
+}
+
+function getGPID(bid) {
+  if (bid.ortb2Imp != null) {
+    if (bid.ortb2Imp.gpid != null & bid.ortb2Imp.gpid != '') {
```

Comment:
i think you're missing a "&"
```suggestion
    if (bid.ortb2Imp.gpid != null && bid.ortb2Imp.gpid != '') {
```

Reply:
Thanks, nice spot